tests: userspace: fail on unexpected faults

stephensmalley opened the pull request:
Previously we were handling any fault during test execution as
a pass condition.  Explicitly indicate when a fault is expected
and fail the test if we encounter an unexpected fault.

Signed-off-by: Stephen Smalley <sds@tycho.nsa.gov>

stephensmalley committed:
tests: userspace: fail on unexpected faults

Previously we were handling any fault during test execution as
a pass condition.  Explicitly indicate when a fault is expected
and fail the test if we encounter an unexpected fault.

Signed-off-by: Stephen Smalley <sds@tycho.nsa.gov>

andrewboie commented on tests/kernel/mem_protect/userspace/src/main.c:
```diff
@@ -20,8 +20,19 @@
+ * been updated.  We'll just make an arbitrary system
+ * call to force one.
+ */
+#define BARRIER() k_sem_give(&expect_fault_sem)
```
can you just use the compiler_barrier() macro defined by toolchain.h?

agross-linaro replied:
i'm not sure the compiler barrier contains what is needed (at least for ARM)

andrewboie replied:
please elaborate.

stephensmalley replied:
We need more than just a compiler barrier here; we need to ensure that the write to expect_fault has completed before the potentially faulting instruction, so we need the processor to honor the memory ordering as well. If this code was running in supervisor mode, I'd just do a __DSB() or __DMB(). The easiest way to do so from usermode seems to be to perform a system call, and repeatedly calling k_sem_give on this semaphore serves that purpose.

andrewboie replied:
gotcha.
more of an "execution barrier" than a "memory barrier".
it's hard to get the compiler to do this :)

tyrantlizard77 reviewed (commented):
This is definitely an improvement.

We might also consider testing the reason code that gets passed to _SysFatalErrorHandler, so that we can assert that the expected type of exception occurred and not something unrelated.

see tests/kernel/fatal for an example of what I mean.

agross-oss reviewed (approved):
looks fine.  maybe something can be done about andrew's comment concerning barrier.

agross-linaro commented:
One additional note on this change.  So right now, if you do any kind of kernel call that ends up calling swap, this will fault inside the fault handler.  We are going to have to fix this to handle any overriding of the Fatal handler.

agross-linaro commented on tests/kernel/mem_protect/userspace/src/main.c:
```diff
@@ -42,7 +53,12 @@ void _SysFatalErrorHandler(unsigned int reason, const NANO_ESF *pEsf)
+		expect_fault = false;
+		BARRIER();
+		ztest_test_pass();
+	}
```
For ARM platforms, we just need to return here right after the ztest_test_pass.  The k_thread_abort will issue a pendsv to swap.  The swap won't happen until we return from exception, so we'll actually fall through to the zassert_unreachable.

andrewboie replied:
no negative, if it is doing that it is a bug.
immediately after we return from the exception, the swap should occur. it should not be executing anything in the application!
See tests/kernel/fatal

stephensmalley replied:
Ah, good point.  We can either return after ztest_test_pass() or move the zassert_unreachable() inside an else clause.

stephensmalley commented:
> We might also consider testing the reason code that gets passed to _SysFatalErrorHandler, so that we can assert that the expected type of exception occurred and not something unrelated.
> 
> see tests/kernel/fatal for an example of what I mean.

We had discussed this earlier, but on ARM we don't get distinct codes and I think you had said to defer it on an overhaul of the fault handling code.  Can certainly add it but it won't be very interesting at least for ARM.

stephensmalley committed:
tests: userspace: fail on unexpected fault reason

Check the fault reason against the expected value.
This is presently architecture-specific, and possibly
reflects a bug on ARM (all faults end up with reason 0,
even though ARM does define a separate value for Oops).

Signed-off-by: Stephen Smalley <sds@tycho.nsa.gov>

stephensmalley commented:
Added the reason checking as a separate commit on top.  Note that this may reflect a bug on ARM, where everything shows up with reason 0.  Latest 4974 passes for me on this and appears to be actually executing the test cases.

agross-linaro commented:
These changes look good.  I tested against 4974 on both NXP frdm-64k  and ARM beetle
